### PR TITLE
agent/app: move most of the code from agent.go to the app package

### PR DESCRIPTION
**Pull request by aaithal**
### Summary

Breaks up agent.go

### Implementation details

* move most of the code from agent.go to the app package
* tests on their way
### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass
- [x] Manual Testing (Can be converted to functional tests if needed)
    - [x] Mismatch in cluster name leads to terminal failure
    - [x] Mismatch in ec2 instance id leads to a state reset when cluster name is not modified 
        - [x] Mismatch in ec2 instance id leads to a state reset when cluster name is modified (#858)
    - [x] Instance type mismatch leads to terminal failure

New tests cover the changes: Yes

### Description for the changelog

None

### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

### CODE REVIEW

File: agent/app/agent.go [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+		// We instantiate our own credentialProvider for use in acs/tcs. This tries
+		// to mimic roughly the way it's instantiated by the SDK for a default
+		// session.
+		credentialProvider: defaults.CredChain(defaults.Config(), defaults.Handlers()),
+	}, nil
+
+}
+
+func (agent *ecsAgent) printVersion() int {
+	version.PrintVersion(agent.dockerClient)
```
Comment (petderek) [inline agent/app/agent.go]: I don't understand this/these abstraction(s).

`agent.printVersion()`  returns the agent version. Why does it need the dockerclient to figure out which version of agent is running?

File: agent/app/agent.go [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+
+	return agent.doStart(containerChangeEventStream, credentialsManager, state, imageManager)
+}
+
+func (agent *ecsAgent) doStart(containerChangeEventStream *eventstream.EventStream,
+	credentialsManager credentials.Manager,
+	state dockerstate.TaskEngineState,
+	imageManager engine.ImageManager) int {
+
+	taskEngine, currentEC2InstanceID, err := agent.newTaskEngine(containerChangeEventStream,
```
Comment (petderek) [inline agent/app/agent.go]: Why does `newTaskEngine` return the ec2 id? Can we attach that to the configuration instead?

File: agent/app/agent.go [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+	err = agent.registerContainerInstance(taskEngine, stateManager, client)
+	if err != nil {
+		if isNonTerminal(err) {
+			return exitcodes.ExitError
+		}
+		return exitcodes.ExitTerminal
+	}
+
+	// Begin listening to the docker daemon and saving changes
+	taskEngine.SetSaver(stateManager)
```
Comment (petderek) [inline agent/app/agent.go]: 'saver?' Why not just call it a manager?

File: agent/app/agent.go [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+
+	deregisterInstanceEventStream := eventstream.NewEventStream(
+		deregisterContainerInstanceEventStreamName, agent.ctx)
+	deregisterInstanceEventStream.StartListening()
+
+	taskHandler := eventhandler.NewTaskHandler()
+
+	agent.startAsyncRoutines(containerChangeEventStream, credentialsManager, imageManager,
+		taskEngine, stateManager, deregisterInstanceEventStream, client, taskHandler)
+	return agent.startACSSession(credentialsManager, taskEngine, stateManager,
```
Comment (petderek) [inline agent/app/agent.go]: This seems odd. Why not start ACS async as well?

File: agent/app/agent.go [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+	containerChangeEventStream.StartListening()
+
+	if !agent.cfg.Checkpoint {
+		log.Info("Checkpointing not enabled; a new container instance will be created each time the agent is run")
+		return engine.NewTaskEngine(agent.cfg, agent.dockerClient, credentialsManager,
+			containerChangeEventStream, imageManager, state), "", nil
+	}
+
+	// We try to set these values by loading the existing state file first
+	var previousCluster, previousEC2InstanceID, previousContainerInstanceArn string
```
Comment (petderek) [inline agent/app/agent.go]: wait -- why would the ec2 instance id change?

File: agent/app/error.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+package app
+
+// nonTerminalError represents a transient error when executing the ECS Agent
+type nonTerminalError struct {
+	error
+}
+
+// isNonTerminal returns true if the error is transient
+func isNonTerminal(err error) bool {
```
Comment (petderek) [inline agent/app/error.go]: Could I suggest naming this (and the type) "transient" instead of "non terminal"? We can then avoid saying things like "not non terminal" in the above code.

File: agent/app/error.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package app
+
```
Comment (petderek) [inline agent/app/error.go]: Naming is inconsistent:
```
./app/error.go
./api/errors.go
```

File: agent/engine/docker_task_engine.go [hunk 8]
```diff
@@ -137,9 +137,9 @@ func (engine *DockerTaskEngine) MarshalJSON() ([]byte, error) {
 // Init initializes a DockerTaskEngine such that it may communicate with docker
 // and operate normally.
 // This function must be called before any other function, except serializing and deserializing, can succeed without error.
-func (engine *DockerTaskEngine) Init() error {
+func (engine *DockerTaskEngine) Init(ctx context.Context) error {
 	// TODO, pass in a a context from main from background so that other things can stop us, not just the tests
-	ctx, cancel := context.WithCancel(context.TODO())
+	derivedCtx, cancel := context.WithCancel(context.TODO())
```
Comment (petderek) [inline agent/engine/docker_task_engine.go]: should you base this off of the parent context (ctx from init?)

File: agent/agent.go [hunk 9]
```diff
@@ -14,310 +14,17 @@
 package main
 
 import (
-	"flag"
-	"fmt"
 	mathrand "math/rand"
```
Comment (aaithal) [inline agent/agent.go]: nope, legacy code. remove it.

File: agent/app/agent.go (hunk 3 above)
Comment (aaithal) [inline agent/app/agent.go]: because naming is hard. The basic state manager is also a saver as it implements the `statemanager.Saver` interface.

File: agent/app/error.go (hunk 7 above)
Comment (aaithal) [inline agent/app/error.go]: I can change it if you insist. But, it's hard trying to keep it consistent across packages.

File: agent/app/error.go (hunk 6 above)
Comment (aaithal) [inline agent/app/error.go]: Sure, i'll change this.

File: agent/engine/docker_task_engine.go (hunk 8 above)
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: That's what I intended to do! Seem to have missed that. I'll make that change.

File: agent/app/agent.go (hunk 1 above)
Comment (aaithal) [inline agent/app/agent.go]: We also print the docker client api version when printing the Agent version. Not sure why we did it initially. But, we can't change it now.

File: agent/app/agent.go (hunk 4 above)
Comment (aaithal) [inline agent/app/agent.go]: If the acs handler stops for whatever reason, we want the Agent to exit. We can start it async, add a wait group or something and wait using that. Instead, we chose to do this.

File: agent/app/agent.go (hunk 5 above)
Comment (aaithal) [inline agent/app/agent.go]: The loaded (previous) ec2 instance id can be different from the current id if launching from an AMI that was created by creating a snapshot of an instance that was registered with ECS.

**Comment by aaithal** [comment]
This is not a WIP anymore. I've added some manual tests in `Testing Done` to build some additional confidence as well and the plan on getting to those soon.

cc @petderek @richardpen @nmeyerhans @samuelkarp

File: agent/app/run.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+	agent, err := newAgent(context.Background(),
+		aws.BoolValue(parsedArgs.BlackholeEC2Metadata),
+		parsedArgs.AcceptInsecureCert)
+	if err != nil {
+		// Failure to initialize either the docker client or the EC2 metadata
+		// service client are non terminal errors as they could be transient
+		return exitcodes.ExitError
+	}
+
+	if *parsedArgs.Version {
```
Comment (aaithal) [inline agent/app/run.go]: We need the `agent` object to print the version. If we could pull out the docker client creation to `run` from `newAgent` we could do what you're saying. But, I don't think it's that much of an improvement.

File: agent/app/license.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"fmt"
+	"os"
+
+	"github.com/aws/amazon-ecs-agent/agent/sighandlers/exitcodes"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
+)
+
+// printLicense prints the Agent's license text
+func printLicense() int {
```
Comment (aaithal) [inline agent/app/license.go]: I did not make it a part of the agent interface as printing the license has got nothing to do with the `agent` object itself.

File: agent/app/agent.go [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+	if err == nil {
+		return nil
+	}
+	log.Errorf("Error re-registering: %v", err)
+	if api.IsInstanceTypeChangedError(err) {
+		log.Criticalf(instanceTypeMismatchErrorFormat, err)
+		return err
+	}
+	if _, ok := err.(utils.AttributeError); ok {
+		log.Criticalf("Instance re-registration attempt with an invalid attribute")
```
Comment (aaithal) [inline agent/app/agent.go]: done

File: agent/app/agent.go [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+		log.Criticalf("Error creating Docker client: %v", err)
+		return nil, err
+	}
+
+	return &ecsAgent{
+		ctx:               ctx,
+		ec2MetadataClient: ec2MetadataClient,
+		cfg:               cfg,
+		dockerClient:      dockerClient,
+		// We instantiate our own credentialProvider for use in acs/tcs. This tries
```
Comment (samuelkarp) [inline agent/app/agent.go]: Is there a good reason to continue using our own `credentialProvider` instead of the default?

File: agent/app/agent.go [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+	}
+
+	if previousCluster != "" {
+		if err := agent.setClusterInConfig(previousCluster); err != nil {
+			return nil, "", err
+		}
+	}
+
+	currentEC2InstanceID := agent.getEC2InstanceID()
+	if previousEC2InstanceID != "" && previousEC2InstanceID != currentEC2InstanceID {
```
Comment (samuelkarp) [inline agent/app/agent.go]: Can you move this check up above the cluster check?  It would be good to reset the task engine if the EC2 instance ID changes regardless of whether the cluster matches the state file.  I think the current behavior where we don't do that is a bug.

File: agent/app/agent.go [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+		// This is for making testing easier as we can mock this
+		agent.saveableOptionFactory.AddSaveable("ContainerInstanceArn",
+			containerInstanceArn),
+		statemanager.AddSaveable("Cluster", cluster),
+		// This is for making testing easier as we can mock this
+		agent.saveableOptionFactory.AddSaveable("EC2InstanceID", savedInstanceID),
+		// The ACSSeqNum field is retained for compatibility with
+		// statemanager.EcsDataVersion 4 and can be removed in the future
+		// with a version bump.
+		statemanager.AddSaveable("ACSSeqNum", 1),
```
Comment (samuelkarp) [inline agent/app/agent.go]: Do we still need this?  We're beyond version 4.

File: agent/app/agent.go (hunk 13 above)
Comment (aaithal) [inline agent/app/agent.go]: I don't think so. Also, I'd prefer to do that in a separate PR.

File: agent/app/agent.go (hunk 14 above)
Comment (aaithal) [inline agent/app/agent.go]: It's a bit odd that we treat these mismatches differently. Just to clarify, ECS Agent will exit with a terminal error if there's a cluster name mismatch. Should that not be the case anymore?

File: agent/app/agent.go (hunk 15 above)
Comment (aaithal) [inline agent/app/agent.go]: Looking at #434, I don't think we need this anymore. I'll get rid of this.

File: agent/app/agent.go (hunk 14 above)
Comment (samuelkarp) [inline agent/app/agent.go]: Yes, I believe that the current behavior is a bug. If the EC2 instance ID is different, it shouldn't matter that the cluster name is mismatched because we'd want to reset the task engine anyway.

Comment (aaithal) [inline agent/app/agent.go]: @samuelkarp This should be fixed now.
